File: 2021/strings/shortest_distance_to_character.py
# Hi, here's your problem today. This problem was recently asked by Uber:

# Given a string s and a character c, find the distance for all characters in 
# the string to the character c in the string s. You can assume that the 
# character c will appear at least once in the string.

# Here's an example and some starter code:
def shortest_dist(s, c):
  result = [float('inf')] * len(s)

  c_idx = float('inf')
  for idx, ch in enumerate(s):
    if ch == c:
      c_idx = idx
      result[idx] = 0
    elif c_idx != float('inf'):
      result[idx] = idx - c_idx

  c_idx = float('inf')
  for idx, ch in reversed(list(enumerate(s))):
    if ch == c:
      c_idx = idx
      result[idx] = 0
    elif c_idx != float('inf'):
      result[idx] = min(result[idx], c_idx - idx)

  return result


print(shortest_dist('helloworld', 'l'))
# [2, 1, 0, 0, 1, 2, 2, 1, 0, 1]

File: strings/shortest_distance_to_character.js
// Hi, here's your problem today. This problem was recently asked by Uber:

// Given a string s and a character c, find the distance for all characters in 
// the string to the character c in the string s. You can assume that the 
// character c will appear at least once in the string.

function shortestDistanceToChar( string, char ) {
	let distanceArr = [];

	// left to right starting at first instance of char and addign distance
	let first = false;
	let distance = 0;
	for ( i = 0; i < string.length; i++ ) {
		if ( string[i] === char ) {
			first = true;
			distance = 0;
			distanceArr.push( 0 );
		} else if ( first ) {
			distance++;
			distanceArr.push( distance );
		} else {
			distanceArr.push( 0 );
		}
	}

	first = false;
	distance = 0;
	for ( i = string.length - 1; i >= 0; i -= 1 ) {
		if ( string[i] === char ) {
			first = true;
			distance = 0;
		} else if ( first ) {
			distance++;
			if ( distanceArr[i] > distance || distanceArr[i] === 0 ) {
				distanceArr[i] = distance;
			}
		}
	}

	return distanceArr;
}

console.log( shortestDistanceToChar( "hello", "l" ) );
console.log( shortestDistanceToChar( "hello world", "l") );